src/ai/backend/manager/api/session.py

```diff
@@ -508,52 +509,70 @@ async def _create(request: web.Request, params: Any) -> web.Response:
                         .where(kernels.c.id == kernel_id)
                     )
                     result = await conn.execute(query)
-                    row = result.first()
-                if row['status'] == KernelStatus.RUNNING:
-                    resp['status'] = 'RUNNING'
-                    for item in row['service_ports']:
-                        response_dict = {
-                            'name': item['name'],
-                            'protocol': item['protocol'],
-                            'ports': item['container_ports'],
-                        }
-                        if 'url_template' in item.keys():
-                            response_dict['url_template'] = item['url_template']
-                        if 'allowed_arguments' in item.keys():
-                            response_dict['allowed_arguments'] = item['allowed_arguments']
-                        if 'allowed_envs' in item.keys():
-                            response_dict['allowed_envs'] = item['allowed_envs']
-                        resp['servicePorts'].append(response_dict)
-                else:
-                    resp['status'] = row['status'].name
+
+                return result.first()
+
+            async def _update_progress(
+                app: web.Application,
+                source: AgentId,
+                event: KernelPullProgressEvent,
+            ) -> None:
+                # update both current and total
+                progress[0] = int(event.current_progress)
+                progress[1] = int(event.total_progress)
+
+            root_ctx.event_dispatcher.subscribe(KernelPullProgressEvent, request.app, _update_progress)
+            kernel_id = resp['sessionId']
+            while True:
+                result = await _get_status(kernel_id)
+                if result is None:
+                    continue
+                if result['status'] == KernelStatus.PREPARING:
+                    await reporter.update(0)
+                if result['status'] == KernelStatus.RUNNING:
+                    break
+                reporter.current_progress = progress[0]
+                reporter.total_progress = progress[1]
+                await reporter.update(0)
+            await asyncio.sleep(0.5)
+
+        task_id = await root_ctx.background_task_manager.start(
+                kernelpullprogress,
+                name='kernel-pull-progress'
+        )
+        resp['background_task'] = str(task_id)
+        resp['sessionName'] = str(params['session_name'])
+        resp['status'] = 'PENDING'
+        resp['servicePorts'] = []
+        resp['created'] = True
+
+        app_ctx.pending_waits.add(current_task)
+        max_wait = params['max_wait_seconds']
+        try:
+            if max_wait > 0:
+                with timeout(max_wait):
+                    await asyncio.sleep(0.5)
+            else:
+                await asyncio.sleep(0.5)
+
+        except asyncio.TimeoutError:
+            resp['status'] = 'TIMEOUT'
+
```

Comment:
This breaks the existing API semantics.
Do we really need to remove these to implement the image pulling progress?